portfolio/src/main/java/com/google/sps/data/LinkedListTimeRangeGroup.java

```diff
@@ -0,0 +1,180 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Models an implementation of the TimeRangeGroup model using LinkedList. This is a slightly
+ * modified version of what hollyyuqizheng wrote for the ArrayList implementation. 
+ */
+public class LinkedListTimeRangeGroup implements TimeRangeGroup {
+
+  // This list of all the time ranges will be sorted by start time ascending
+  // in add and delete methods. The time ranges stored in this list are
+  // pair-wise disjoint at any moment.
+  public List<TimeRange> allTimeRanges;
+
+  /**
+   * Adds all the input time ranges to the list of all time ranges. Also sorts the list of all
+   * ranges in the constructor.
+   */
+  public LinkedListTimeRangeGroup(Collection<TimeRange> timeRanges) {
+    allTimeRanges = new LinkedList<TimeRange>();
+    timeRanges.forEach(
+        (range) -> {
+          addTimeRange(range);
+        });
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    // If the original allTimeRanges list is empty,
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+
+    List<TimeRange> newTimeRanges = new LinkedList<TimeRange>();
+
+    // This variable represents the latest time range previously exmamined.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    for (int i = 0; i < allTimeRanges.size(); i++) {
+      TimeRange currentRange = allTimeRanges.get(i);
```

Comment:
Doing this loop by referring to an index was O(n) with the arraylist but is now O(n^2) with a linked list.